File: sdk/go/preset_engine.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

// Parameter represents a complex UI-aware parameter
type Parameter struct {
	Value   interface{} `json:"value"`
	UI      string      `json:"ui,omitempty"`
	Style   string      `json:"style,omitempty"`
	Min     float64     `json:"min,omitempty"`
	Max     float64     `json:"max,omitempty"`
	Options []string    `json:"options,omitempty"`
}

// Effect is a generic map to hold any effect configuration
type Effect map[string]interface{}

// Chain is a list of effects
type Chain []Effect

// NewChain creates a new empty chain
func NewChain() *Chain {
	return &Chain{}
}

// Add appends an effect to the chain
func (c *Chain) Add(e Effect) {
	*c = append(*c, e)
}

// ToJSON returns the JSON string representation
func (c *Chain) ToJSON() (string, error) {
	b, err := json.MarshalIndent(c, "", "  ")
	return string(b), err
}

// --- Helpers ---

func Gain(db float64) Effect {
	return Effect{
		"type":    "Gain",
		"gain_db": db,
	}
}

func GainUI(db float64) Effect {
	return Effect{
		"type": "Gain",
		"gain_db": Parameter{
			Value: db,
			UI:    "Slider",
			Style: "LinearVertical",
			Min:   -60,
			Max:   12,
		},
	}
}

func Filter(mode string, freq float64) Effect {
	return Effect{
		"type":      "Filter",
		"type_mode": mode, // Note: 'type' is reserved for effect type, so we might need to handle this key collision in Go maps carefully or use a struct.
		// Actually, for the plugin 'type' property inside Filter, we can just use "mode" or "filter_type" if we updated the C++ side, 
		// but the C++ side expects "type" property for the filter mode too?
		// Let's check FilterEffect.h. It looks for "type". 
		// In JSON: { "type": "Filter", "type": "LowPass" } is invalid JSON (duplicate key).
		// Wait, my previous JSON example had this issue and I fixed it by changing the inner one to "mode".
		// I need to ensure C++ FilterEffect supports "mode".
		"frequency": freq,
	}
}

// We need to verify FilterEffect.h supports "mode". 
// I recall adding "mode" support for LadderFilter, but standard FilterEffect might still be looking for "type".
// I should check FilterEffect.h and update it if necessary to support "mode" to avoid JSON key collision.

File: sdk/python/preset_engine.py
import json
import yaml

class Effect:
    def __init__(self, type_name, **kwargs):
        self.type = type_name
        self.params = kwargs

    def to_dict(self):
        d = {"type": self.type}
        for k, v in self.params.items():
            if isinstance(v, Parameter):
                d[k] = v.to_dict()
            else:
                d[k] = v
        return d

class Parameter:
    def __init__(self, value, ui=None, style=None, min_val=None, max_val=None, options=None):
        self.value = value
        self.ui = ui
        self.style = style
        self.min = min_val
        self.max = max_val
        self.options = options

    def to_dict(self):
        d = {"value": self.value}
        if self.ui: d["ui"] = self.ui
        if self.style: d["style"] = self.style
        if self.min is not None: d["min"] = self.min
        if self.max is not None: d["max"] = self.max
        if self.options: d["options"] = self.options
        return d

class Chain:
    def __init__(self):
        self.effects = []

    def add(self, effect):
        self.effects.append(effect)
        return self

    def to_json(self):
        return json.dumps([e.to_dict() for e in self.effects], indent=2)

    def to_yaml(self):
        return yaml.dump([e.to_dict() for e in self.effects], sort_keys=False)

# --- Effect Wrappers ---

def Gain(db=0.0, ui=False):
    if ui:
        return Effect("Gain", gain_db=Parameter(db, ui="Slider", style="Linear", min=-60, max=12))
    return Effect("Gain", gain_db=db)

def Filter(mode="LowPass", freq=1000.0, q=0.7, ui=False):
    if ui:
        return Effect("Filter", 
                      type=Parameter(mode, ui="ComboBox", options=["LowPass", "HighPass", "BandPass"]),
                      frequency=Parameter(freq, ui="Slider", style="Rotary", min=20, max=20000),
                      q=Parameter(q, ui="Slider", min=0.1, max=10))
    return Effect("Filter", type=mode, frequency=freq, q=q)

def Reverb(size=0.5, wet=0.33):
    return Effect("Reverb", room_size=size, wet=wet)
